transfer_nlp/plugins/config.py

```diff
@@ -157,50 +156,162 @@ def __init__(self, experiment: Union[str, Path, Dict], **env):
+                f'Object of class {object_dict["_name"]}, which is not registered. see transfer_nlp.config.register_plugin for more information')
+
+        spec = inspect.getfullargspec(clazz.__init__)
+        params = {}
+        param2config_key = {}
+        named_params = {p: pv for p, pv in object_dict.items() if p != '_name'}
+        default_params = {p: pv for p, pv in zip(reversed(spec.args), reversed(spec.defaults))} if spec.defaults else {}
+
+        for arg in spec.args[1:]:
+
+            if arg == 'experiment_config':
+                params[arg] = self
+                param2config_key[arg] = arg
+
+            elif arg in named_params:
+                value = named_params[arg]
+
+                if isinstance(value, dict):
+                    if '_name' in value:
+                        value = self._do_recursive_build(object_key=arg, object_dict=value, default_params_mode=default_params_mode)
+                    else:
+                        for item in value:
+                            if isinstance(value[item], dict):
+                                value[item] = self._do_recursive_build(object_key=item, object_dict=value[item], default_params_mode=default_params_mode)
+                            else:   # value[item] is either an object defined in a dictionary, or it's an already built object
+                                logger.info(f"{item} is already configured")
+                elif isinstance(value, str) and value[0] == '$':
+                    if value[1:] in self.experiment:
+                        logger.info(f"Using the object {value}, already instantiated")
+                        value = self.experiment[value[1:]]
+                    else:
+                        logger.info(f"{value} not configured yet, will be configured in next iteration")
+                else:
+                    logger.info(f"Using value {arg} / {named_params[arg]} from the config file")
+
+                params[arg] = value
+                param2config_key[arg] = value
+
+            # For values that are not in named_params, we look first at the experiment dict, then at the defaults parameters
+            elif arg in self.experiment:
+                params[arg] = self.experiment[arg]
+                param2config_key[arg] = arg
+
+            elif default_params_mode == 1 and arg not in self.experiment and arg in default_params and default_params[arg] is not None:
+                params[arg] = default_params[arg]
+                param2config_key[arg] = None
+            elif default_params_mode == 2 and arg in default_params:
+                params[arg] = default_params[arg]
+                param2config_key[arg] = None
+            else:
+                raise ValueError(f"{arg} is not a parameter from the {class_name} class")
+
+        if len(params) == len(spec.args) - 1:
+
+            self.factories[object_key] = PluginFactory(cls=clazz, param2config_key=param2config_key, **params)
```

Comment:
i think there's still a problem with regards to the factory keys. the factory keys should be qualified by the parent key for nested configs, e.g. if in RegistryTest.test_recursive_definition() i add

```
for k,v in e.factories.items():
   print(k,v)
```

the keys are not qualified. this could lead to collisions if the same class is nested in multiple places in the config